File: apps/comms_web/lib/comms_web/presenters/identity.ex
defmodule CommsWeb.Presenters.Identity do
  @moduledoc false

  alias CommsCore.Accounts.{
    DeviceView,
    DirectoryPersonView,
    RetainedSenderLabelView,
    SessionView,
    UserView
  }

  alias CommsCore.Administration.InvitedIdentityReceipt

  def user(%UserView{} = user) do
    %{
      id: user.id,
      tenant_id: user.tenant_id,
      display_name: user.display_name,
      email: if(user.account_type == :guest, do: nil, else: user.email),
      account_type: user.account_type,
      access_scope: user.access_scope,
      role: user.role,
      status: user.status,
      version: user.version
    }
  end

  def user(%InvitedIdentityReceipt{} = user) do
    %{
      id: user.id,
      tenant_id: user.tenant_id,
      display_name: user.display_name,
      email: user.email,
      account_type: user.account_type,
      role: user.role,
      status: user.status,
      version: user.version
    }
  end

  def identity_user(%UserView{} = user) do
    Map.merge(user(user), %{
      platform_role: user.platform_role,
      platform_role_expires_at: user.platform_role_expires_at
    })
  end

  def guest_identity(%UserView{} = user) do
    %{
      id: user.id,
      tenant_id: user.tenant_id,
      display_name: user.display_name,
      account_type: user.account_type,
      access_scope: user.access_scope,
      role: user.role,
      status: user.status,
      guest_expires_at: user.guest_expires_at
    }
  end

  def admin_user(%UserView{} = user), do: identity_user(user)

  def directory_person(%DirectoryPersonView{} = person) do
    %{id: person.id, display_name: person.display_name}
  end

  def retained_sender_label(%RetainedSenderLabelView{} = label) do
    %{id: label.id, display_name: label.display_name, redacted: label.redacted}
  end

  def device(%DeviceView{} = device) do
    %{
      id: device.id,
      user_id: device.user_id,
      name: device.name,
      platform: device.platform,
      last_seen_at: device.last_seen_at,
      revoked_at: device.revoked_at
    }
  end

  def session(%SessionView{} = session) do
    Map.merge(
      %{
        id: session.id,
        user_id: session.user_id,
        device_id: session.device_id,
        expires_at: session.expires_at,
        last_used_at: session.last_used_at,
        revoked_at: session.revoked_at,
        inserted_at: session.inserted_at
      },
      %{
        platform_role: session.platform_role,
        platform_role_expires_at: session.platform_role_expires_at
      }
    )
  end
end

File: clients/web/src/types/identity.ts
import type { Conversation } from "./messaging";

export interface Tenant {
  id: string;
  name: string;
  slug: string;
  status: string;
}

export type UserRole =
  | "member"
  | "moderator"
  | "admin"
  | "compliance_admin"
  | "security_admin"
  | "owner";

export type PlatformRole = "platform_operator" | "support_operator" | "security_operator";

export interface User {
  id: string;
  tenant_id: string;
  display_name: string;
  email?: string | null;
  account_type?: "human" | "service" | "guest";
  role: UserRole;
  platform_role?: PlatformRole | null;
  platform_role_expires_at?: string | null;
  status: string;
  version?: number;
}

export interface DirectoryPerson {
  id: string;
  display_name: string;
}

export interface RetainedSenderLabel {
  id: string;
  display_name: string;
  redacted: boolean;
}

export interface DirectoryPeoplePage {
  data: DirectoryPerson[];
  page: {
    next_cursor: string | null;
  };
}

export interface Device {
  id: string;
  user_id: string;
  name: string;
  platform: string;
  last_seen_at?: string | null;
  revoked_at?: string | null;
}

export interface Session {
  access_token: string;
  refresh_token: string;
  token_type: string;
  expires_in: number;
  received_at?: number;
  tenant: Tenant;
  user: User;
  device: Device;
}

export interface GuestLink {
  id: string;
  conversation_id: string;
  expires_at: string;
  max_uses: number;
  use_count: number;
  conversion_enabled?: boolean;
  email_hint?: string | null;
  status: "active" | "expired" | "revoked" | "exhausted";
  revoked_at?: string | null;
  version?: number;
  share_url?: string;
}

export interface GuestCapabilities {
  allow_audio_calls: boolean;
  allow_video_calls: boolean;
  /**
   * Missing during a rolling upgrade is treated as disabled by the client.
   * Account conversion is available only on a host-preauthorized, single-use
   * link and still requires the separately delivered one-time code.
   */
  conversion_enabled?: boolean;
  /**
   * Instant-room guests can upgrade the identity already in the room without
   * a host-issued verifier. This is intentionally separate from the stricter
   * pre-authorized conversion used by ordinary guest invitations.
   */
  self_service_conversion?: boolean;
  /** Masked display hint only; the guest must enter the full authorized email. */
  email_hint?: string | null;
}

export interface GuestLinkPreview {
  room_title: string;
  expires_at: string;
  conversion_enabled: boolean;
  email_hint: string | null;
}

export interface GuestSession extends Session {
  conversation: Conversation;
  capabilities: GuestCapabilities;
  admission?: {
    guest_link_id: string;
    expires_at: string;
  };
  instant_room?: InstantRoom;
  /** The exact public URL returned by the server. Never reconstructed client-side. */
  share_url?: string;
}

export type InstantRoomOwnerKind = "guest" | "registered";

export type InstantRoomStatus = "active" | "idle" | "expired" | "revoked";

export interface InstantRoom {
  id: string;
  conversation_id: string;
  owner_user_id: string;
  status: InstantRoomStatus;
  owner_kind: InstantRoomOwnerKind;
  participant_limit: number;
  idle_since: string | null;
  expires_at: string | null;
  inserted_at: string;
  updated_at: string;
}

export interface InstantRoomPreview {
  room_title: string;
  status: InstantRoomStatus;
  expires_at: string | null;
  participant_limit: number;
}

export interface InstantRoomResult {
  room: InstantRoom;
  conversation: Conversation;
  /** The exact public URL returned by the server. */
  share_url?: string;
  /** Present only when an anonymous visitor was admitted as the room creator. */
  guest_session?: GuestSession;
}

export interface SocketHandoff {
  ticket: string;
  expires_in: number;
}

export interface GuestAccountConversionResult {
  session: Session;
  conversation: Conversation;
  socket_handoff?: SocketHandoff;
}

export interface MeResponse {
  tenant: Tenant;
  user: User;
  device: Device;
  capabilities: UserCapabilities;
}

export interface UserCapabilities {
  allow_audio_calls: boolean;
  allow_video_calls: boolean;
  /**
   * Optional because a server that predates the immersive increment omits it,
   * and an absent switch must read as "not eligible" rather than as "true".
   */
  allow_immersive_mode?: boolean;
  allow_public_channels: boolean;
  message_edit_window_seconds: number;
  max_attachment_bytes: number;
}
